src/AutomaticDifferentiation/Enzyme/SparseWrapper.hpp

```diff
@@ -14,163 +14,168 @@ namespace GridKit
-     *
-     * @tparam ScalarT - scalar data type
-     */
-    template <typename ScalarT>
-    __attribute__((always_inline)) static void ident_store(ScalarT, size_t, size_t)
-    {
-      assert(0 && "should never load");
-    }
-
-    /**
-     * @brief Enzyme identity load
-     *
-     * @tparam ScalarT - scalar data type
-     */
-    template <typename ScalarT>
-    __attribute__((always_inline)) static ScalarT ident_load(size_t idx, size_t i)
-    {
-      idx /= sizeof(ScalarT);
-      return (ScalarT) (idx == i);
-    }
-
-    /**
-     * @brief Function that computes the Jacobian via automatic differentiation
-     *
-     * @tparam ModelT - model type
-     * @tparam ScalarT - scalar data type
-     * @tparam IdxT    - matrix index data type
-     */
-    template <typename ModelT, class ScalarT, typename IdxT>
-    __attribute__((noinline)) void EnzymeSparseModelJacobian(ModelT* model, size_t n, ScalarT* input, GridKit::LinearAlgebra::COO_Matrix<ScalarT, IdxT>& jac)
-    {
-      std::vector<Triple<ScalarT>> triplets;
-      for (size_t i = 0; i < n; i++)
+      /**
+       * @brief Enzyme fwddiff template
+       *
+       * @tparam T - return type
+       * @tparam ModelT - model type
+       */
+      template <typename T, typename... ModelT>
+      extern T __enzyme_fwddiff(void*, ModelT...) noexcept;
+
+      /**
+       * @brief Enzyme todense template
+       *
+       * @tparam T - return type
+       * @tparam ModelT - model type
+       */
+      template <typename T, typename... ModelT>
+      extern T __enzyme_todense(ModelT...) noexcept;
+
+      /**
+       * @brief Enzyme sparse storage in triplet format
+       *
+       * @tparam ScalarT - scalar data type
+       */
+      template <typename ScalarT>
+      struct Triple
+      {
+        size_t  row;
+        size_t  col;
+        ScalarT val;
+        Triple(Triple&&) = default;
+
+        Triple(size_t row, size_t col, ScalarT val)
+          : row(row),
+            col(col),
+            val(val)
+        {
+        }
+      };
+
+      /**
+       * @brief Enzyme sparse accumulation for float
+       *
+       */
+      [[maybe_unused]] __attribute__((enzyme_sparse_accumulate)) static void inner_storeflt(size_t row, size_t col, float val, std::vector<Triple<float>>& triplets)
+      {
+        triplets.emplace_back(row, col, val);
+      }
+
+      /**
+       * @brief Enzyme sparse accumulation for double
+       *
+       */
+      [[maybe_unused]] __attribute__((enzyme_sparse_accumulate)) static void inner_storedbl(size_t row, size_t col, double val, std::vector<Triple<double>>& triplets)
+      {
+        triplets.emplace_back(row, col, val);
+      }
+
+      /**
+       * @brief Enzyme sparse store
+       *
+       * @tparam ScalarT - scalar data type
+       */
+      template <typename ScalarT>
+      __attribute__((always_inline)) static void sparse_store(ScalarT val, size_t idx, size_t i, std::vector<Triple<ScalarT>>& triplets)
```

Comment:
just a minor note: we can use `[[gnu::always_inline]]` for consistency with the other attributes (`[[maybe_unused]]`). this isn't strictly necessary, though. see: https://clang.llvm.org/docs/AttributeReference.html#always-inline-force-inline